### ISSUE TITLE
Configure an optional <file> path

### ISSUE THREAD

**Issue by jigarius** [squizlabs/PHP_CodeSniffer#3560]
## The problem

- I'm trying to define a generic coding standard for my team, which will be used across multiple Drupal projects.
- A Drupal project can have 2 possible names for the document root:
  - web (standard)
  - docroot (non-standard, but used by companies like Acquia)
- No matter the docroot name, I'd like my `phpcs.xml.dist` to cover the following directories:
  - [document-root]/modules/custom
  - [document-root]/profiles/custom
  - [document-root]/themes/custom
- Say, the project uses `docroot` as the `[document-root]`.
  - When I run `phpcs`, it complains about `web/modules/custom` not existing, which is normal because this project uses the directory `docroot` instead.

## Extract of phpcs.xml.dist

```xml
<ruleset name="Drupal - Base" namespace="Drupal\CS\Standard">
    <file>web/modules/custom</file>
    <file>docroot/modules/custom</file>
    <file>web/themes/custom</file>
    <file>docroot/themes/custom</file>
    <file>web/profiles/custom</file>
    <file>docroot/profiles/custom</file>
</ruleset>
```

## Question

- How do I tell `phpcs` that it is okay if one of the `<file>` elements doesn't actually exist. If it doesn't exist, simply ignore it.

**Comment by jrfnl** [comment]
Well, that would require a slightly different setup.

IIRC, the `file>` directive does not allow for regular expressions, however, the `<include-pattern>` directive does. Except that `<include-pattern>` can only be used within `<rule ...>` nodes, so you would need to add it to every `<rule>` or use two rulesets.

In code:
```xml
<ruleset name="Drupal - Base" namespace="Drupal\CS\Standard">
    <rule ref="PSR12">
        <include-pattern>(web|docroot)/modules/custom/*</include-pattern>
        <include-pattern>(web|docroot)/themes/custom/*</include-pattern>
        <include-pattern>(web|docroot)//profiles/custom/*</include-pattern>
    </rule>

    <rule ref="AnotherRule">
        <include-pattern>(web|docroot)/modules/custom/*</include-pattern>
        <include-pattern>(web|docroot)/themes/custom/*</include-pattern>
        <include-pattern>(web|docroot)//profiles/custom/*</include-pattern>
    </rule>
</ruleset>
```



Or using two rulesets:
```xml
<ruleset name="RulesetA" namespace="Drupal\CS\Standard">
    <rule ref="..."/>
    <rule ref="..."/>
</ruleset>
```

```xml
<ruleset name="RulesetB" namespace="Drupal\CS\Standard">
    <rule ref="RulesetA">
        <include-pattern>(web|docroot)/modules/custom/*</include-pattern>
        <include-pattern>(web|docroot)/themes/custom/*</include-pattern>
        <include-pattern>(web|docroot)//profiles/custom/*</include-pattern>
    </rule>
</ruleset>
```

**Comment by jrfnl** [comment]
@jigarius Just checking: did my previous answer help ? Can this issue be closed now ?

**Comment by jigarius** [comment]
I haven't had a chance to try it yet. I will close the issue for now. Need be, I will reopen it. Thanks for your help.

**Comment by jigarius** [comment]
Here's what I've done as per my understanding of the instructions provided by @jrfnl:

- I created a ruleset named, say, `DrupalCustom`.
- I created another ruleset named `ProjectCustom`.

```yaml
<?xml version="1.0"?>
<ruleset namespace="Drupal\CS\Standard">

  <!-- See https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-ruleset -->

  <description>A wrapper for the "DrupalSymetris" coding standard.</description>

  <rule ref="DrupalCustom">
    <include-pattern>(web|docroot)/(modules|themes|profiles)/custom</include-pattern>
  </rule>

</ruleset>
```

This gives me an error saying, at least one `<file>` entry must be defined. So, I defined it with a `<file>.</file>`. Here are the problems I face: The files I am concerned about are only under the `include-pattern` I have defined. However, `phpcs` now scans things like `web/core` and `web/libraries` which have thousands of files.

So, I am tempted to use a solution that I had thought of: Writing my standard file assuming the existence of `web/(modules|themes|profiles)/custom` directories and then making sure they exist. If the Drupal project uses a document root of `docroot` then I'll create a symlink named `web` just to make `phpcs` happy. Not the best solution, but not much I can do it seems.

**Comment by jrfnl** [comment]
@jigarius You can use `<exclude-pattern>` at the top-level, so if you know of typical directories, which shouldn't be scanned, add one or more `<exclude-pattern>` directives. Note: these can also contain the PHPCS-flavoured regexes.
